File: DSL/de.unibonn.simpleml/src/main/kotlin/de/unibonn/simpleml/utils/IdManager.kt
package de.unibonn.simpleml.utils

import de.unibonn.simpleml.simpleML.SmlAbstractObject
import java.util.WeakHashMap

/**
 * The ID of an object.
 */
@JvmInline
@Suppress("unused")
value class Id<out T : Any>(val value: Int)

/**
 * Handles the mapping of objects, usually [SmlAbstractObject]s in the Simple-ML AST, to their IDs.
 */
class IdManager<UPPER : Any> {

    /**
     * Maps an object to an ID.
     */
    private val objToId = WeakHashMap<UPPER, Id<UPPER>>()

    /**
     * Maps an ID to an object.
     */
    private val idToObj = WeakHashMap<Id<UPPER>, UPPER>()

    /**
     * The next available ID.
     */
    private var nextId = 0

    /**
     * Assigns the next available ID to the given object unless it already has one and returns the ID for this object.
     */
    fun <T : UPPER> assignIdIfAbsent(obj: T): Id<T> {
        if (obj !in objToId) {
            val id = nextId<T>()
            objToId[obj] = id
            idToObj[id] = obj
        }

        // We only write to objToId here and can be sure the stored ID has the correct type
        @Suppress("UNCHECKED_CAST")
        return objToId[obj]!! as Id<T>
    }

    /**
     * Returns the next available ID.
     */
    private fun <T : Any> nextId() = Id<T>(nextId++)

    /**
     * Returns the object with the given ID or `null` if the ID was not assigned yet.
     */
    fun getObjectById(id: Id<*>) = idToObj[id]

    /**
     * Checks if the given object already has an ID.
     */
    fun knowsObject(obj: Any) = obj in objToId

    /**
     * Check if the given ID has already been assigned to some object.
     */
    fun knowsId(id: Id<*>) = id in idToObj

    /**
     * Removes all mappings between object and ID and resets the counter.
     */
    fun reset() {
        objToId.clear()
        idToObj.clear()
        nextId = 0
    }
}

File: src/language/helpers/idManager.ts
/**
 * Handles the mapping of objects, usually nodes of a Safe-DS AST, to their IDs.
 */
export class IdManager<T extends WeakKey> {
    /**
     * Maps an object to an ID.
     */
    private objToId: WeakMap<T, Id> = new WeakMap();

    /**
     * The next available ID.
     */
    private nextId = 0;

    /**
     * Assigns the next available ID to the given object unless it already has one and returns the ID for this object.
     */
    assignId(obj: T): Id {
        if (!this.objToId.has(obj)) {
            this.objToId.set(obj, this.nextId++);
        }
        return this.objToId.get(obj)!;
    }

    /**
     * Removes all mappings between object and ID and resets the counter.
     */
    reset() {
        this.objToId = new WeakMap();
        this.nextId = 0;
    }
}

export type Id = number;
